Data/HashMap/Base.hs

```diff
@@ -708,36 +988,107 @@ delete k0 m0 = go h0 k0 0 m0
                 | otherwise -> Collision h (A.delete v i)
             Nothing -> t
         | otherwise = t
-{-# INLINABLE delete #-}
+{-# INLINABLE delete' #-}
+
+-- Delete optimized for the case when we know the key is in the map and there is
+-- no collision.
+--
+-- It is only valid to call this when the key exists in the map and you know the
+-- hash collision position if there was one. This information can be obtained
+-- from 'lookupRecordCollision'. If there is no collision pass (-1) as collPos.
+--
+-- We can skip:
+--  - the key equality check on the leaf, if we reach a leaf it must be the key
+deleteKeyExists :: Int -> Hash -> k -> HashMap k v -> HashMap k v
+deleteKeyExists !collPos0 !h0 !k0 !m0 = dkego collPos0 h0 k0 0 m0
+{-# NOINLINE deleteKeyExists #-}
+
+dkego :: Int -> Hash -> k -> Int -> HashMap k v -> HashMap k v
```

Comment:
Any reason for making this top level instead of a in the `where` clause of deleteKeyExists?

Reply:
I'm not sure. I have some questions about whether certain `letrec`s end up allocating closures. I'll probably have to benchmark or something.